File: c/leetcode/876. Middle of the Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode *middleNode(struct ListNode *head)
{
  struct ListNode *late = head;
  struct ListNode *early = head->next;

  while (early)
  {
    early = early->next ? early->next->next : NULL;
    late = late->next;
  }

  return late;
}

File: javascript/leetcode/876. Middle of the Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
  let late = head;
  let early = head.next;

  while (early) {
    early = early.next ? early.next.next : null;
    late = late.next;
  }

  return late;
};
